### PR TITLE
feat: Add IBM Docling integration for enhanced document processing

### LINKED ISSUE #255: Enhancement: Integrate IBM Docling for Advanced Document Processing
## Summary

Replace current document processors with IBM Docling to significantly enhance document ingestion capabilities, including superior table extraction, layout analysis, reading order detection, and support for additional file formats.

### Current Implementation

RAG Modulo currently uses custom processors for document ingestion:
- **PDF Processing**: PyMuPDF (`pymupdf`) - `backend/rag_solution/data_ingestion/pdf_processor.py`
- **Word Processing**: python-docx - `backend/rag_solution/data_ingestion/word_processor.py`
- **Excel Processing**: openpyxl - `backend/rag_solution/data_ingestion/excel_processor.py`
- **Text Processing**: Custom text processor - `backend/rag_solution/data_ingestion/txt_processor.py`

**Current Limitations:**
- Limited to 4 file formats (.pdf, .docx, .xlsx, .txt)
- Basic table extraction using PyMuPDF's built-in methods
- No layout analysis or reading order detection
- No formula/code detection in PDFs
- Basic metadata extraction
- No PowerPoint, HTML, or image format support
- Manual processor management per file type

### IBM Docling Overview

[Docling](https://github.com/docling-project/docling) is IBM's open-source, MIT-licensed document processing toolkit with:

**Key Features:**
- Support for PDF, DOCX, PPTX, XLSX, HTML, images, audio (WAV/MP3)
- AI-powered layout analysis using DocLayNet model
- Advanced table structure recognition with TableFormer model
- Reading order detection for multi-column documents
- Formula and code extraction from PDFs
- Image classification
- Unified DoclingDocument representation
- Export to Markdown, HTML, JSON
- Pre-built integrations with LangChain, LlamaIndex

**Recent Developments (2025):**
- Granite-Docling-258M: Ultra-compact VLM (258M parameters) for one-shot document processing
- 37,000+ GitHub stars
- Hosted by LF AI & Data Foundation
- Active development by IBM Research

### High Impact Improvements

1. **Enhanced Table Extraction**
   - TableFormer AI model significantly outperforms PyMuPDF for complex tables
   - Better handling of merged cells, nested tables, and irregular structures
   - **Impact**: Critical for enterprise documents with financial data, reports

2. **Reading Order Detection**
   - AI-powered layout analysis determines correct reading flow
   - Essential for multi-column documents, scientific papers, magazines
   - **Impact**: Improves RAG search quality by preserving document semantics

3. **Format Expansion**
   - Add PPTX support (presentations)
   - Add HTML support (web content)
   - Add image format support (PNG, JPEG, TIFF with OCR)
   - **Impact**: Expands RAG Modulo's document ingestion capabilities without custom processors

4. **Reduced Maintenance**
   - Single library replaces 4+ custom processors
   - IBM-maintained with active development
   - **Impact**: Reduced technical debt, faster feature adoption

5. **Better Structure Preservation**
   - Layout-aware extraction maintains document hierarchy
   - Preserves headings, sections, lists, code blocks
   - **Impact**: Improved context for RAG retrieval

### Moderate Impact Improvements

- Formula/code detection useful for technical/scientific documents
- Image classification for better image chunk metadata
- Markdown export for document preview/debugging
- Alignment with existing WatsonX integration strategy

### Option 1: Full Replacement (Recommended)

Create a unified Docling adapter that handles all document types:

**New File**: \`backend/rag_solution/data_ingestion/docling_processor.py\`

\`\`\`python
from docling import DocumentConverter
from rag_solution.data_ingestion.base_processor import BaseProcessor

class DoclingProcessor(BaseProcessor):
    """Unified document processor using IBM Docling."""
    
    def __init__(self, settings: Settings):
        super().__init__(settings)
        self.converter = DocumentConverter()
    
    async def process(self, file_path: str, document_id: str) -> AsyncIterator[Document]:
        """Process any document type using Docling."""
        result = self.converter.convert(file_path)
        
        # Convert DoclingDocument to RAG Modulo Document format
        for chunk in self._convert_to_chunks(result, document_id):
            yield chunk
    
    def _convert_to_chunks(self, docling_doc, document_id: str) -> list[Document]:
        """Convert Docling's DoclingDocument to RAG Modulo Document format."""
        # Preserve metadata, layout information, table structures
        # Apply existing chunking strategies
        # Maintain compatibility with embedding pipeline
        pass
\`\`\`

**Update**: \`backend/rag_solution/data_ingestion/document_processor.py\`

\`\`\`python
self.processors: dict[str, BaseProcessor] = {
    ".pdf": DoclingProcessor(settings),    # Replace PyMuPDF
    ".docx": DoclingProcessor(settings),   # Replace python-docx
    ".pptx": DoclingProcessor(settings),   # NEW FORMAT
    ".html": DoclingProcessor(settings),   # NEW FORMAT
    ".png": DoclingProcessor(settings),    # NEW FORMAT
    ".jpg": DoclingProcessor(settings),    # NEW FORMAT
    ".txt": TxtProcessor(settings),        # Keep for simplicity
    ".xlsx": ExcelProcessor(settings),     # Keep for simplicity
}
\`\`\`

### Option 2: Hybrid Approach (Lower Risk)

- Keep current processors for simple formats (.txt, .xlsx)
- Use Docling for complex formats (.pdf, .docx, .pptx, .html)
- Gradual migration with feature flag

### Phase 1: Setup & Infrastructure
- [ ] Add \`docling\` dependency to \`backend/pyproject.toml\`
- [ ] Create \`DoclingProcessor\` class in \`backend/rag_solution/data_ingestion/\`
- [ ] Implement DoclingDocument → Document adapter
- [ ] Add feature flag for Docling vs legacy processors

### Phase 2: Core Integration
- [ ] Integrate DoclingProcessor with PDF files
- [ ] Update \`DocumentProcessor\` to route to appropriate processor
- [ ] Ensure compatibility with existing embedding pipeline (\`DocumentStore._embed_documents_batch()\`)
- [ ] Preserve existing chunking strategies integration

### Phase 3: Testing & Validation
- [ ] Unit tests: Compare Docling output vs current processors
- [ ] Integration tests: Full ingestion pipeline with Docling
- [ ] Performance benchmarks: Processing speed, memory usage
- [ ] Quality validation: Table extraction accuracy, reading order correctness
- [ ] Test on representative document corpus

### Phase 4: Format Expansion
- [ ] Add PPTX support
- [ ] Add HTML support  
- [ ] Add image format support (PNG, JPEG)
- [ ] Update API documentation with new supported formats

### Phase 5: Migration & Rollout
- [ ] Gradual rollout with feature flag
- [ ] Monitor performance metrics
- [ ] Deprecate old processors
- [ ] Update documentation

### Dependencies
\`\`\`toml
# backend/pyproject.toml
[tool.poetry.dependencies]
docling = "^2.0.0"
\`\`\`

### Performance
- Docling runs efficiently on commodity hardware (no GPU required)
- AI models (DocLayNet, TableFormer) have small resource footprint
- Compatible with existing async/batch processing architecture

### Compatibility
- Maintains compatibility with existing \`Document\`/\`DocumentChunk\` schema
- Works with current embedding generation pipeline
- Preserves chunking strategies (simple, semantic, token-based)

### Risks & Mitigations

**Risks:**
1. New dependency introduces potential instability
   - **Mitigation**: MIT license, IBM-backed, active development, 37K+ stars
2. Performance impact of AI models
   - **Mitigation**: Benchmark before rollout, feature flag for rollback
3. Breaking changes to document processing behavior
   - **Mitigation**: Side-by-side testing, gradual migration
4. Migration effort for existing ingested documents
   - **Mitigation**: Optional re-ingestion, version document processor in metadata

**Mitigation Strategy:**
- Start with hybrid approach (PDF only via Docling)
- Feature flag to toggle between processors
- Comprehensive benchmarking on production document corpus
- Fallback to legacy processors if Docling fails

### Quantitative
- Table extraction accuracy improvement (target: >30%)
- Reading order correctness for multi-column documents (target: >90%)
- Support for 3+ new file formats (PPTX, HTML, images)
- Performance: Processing time within 20% of current implementation

### Qualitative
- Improved RAG search quality for documents with complex layouts
- Reduced maintenance burden (fewer custom processors)
- Better document structure preservation in chunks

## References

- [Docling GitHub Repository](https://github.com/docling-project/docling) (37K+ stars)
- [Docling Documentation](https://docling-project.github.io/docling/)
- [IBM Research: Docling Announcement](https://research.ibm.com/blog/docling-generative-AI)
- [Granite-Docling-258M Model](https://huggingface.co/ibm-granite/granite-docling-258M)
- [IBM Announcement: Granite-Docling End-to-End](https://www.ibm.com/new/announcements/granite-docling-end-to-end-document-conversion)

## Related Files

**Current Implementation:**
- \`backend/rag_solution/data_ingestion/document_processor.py\` - Main processor orchestrator
- \`backend/rag_solution/data_ingestion/pdf_processor.py\` - PDF processing (to be replaced)
- \`backend/rag_solution/data_ingestion/word_processor.py\` - Word processing (to be replaced)
- \`backend/rag_solution/data_ingestion/excel_processor.py\` - Excel processing (keep)
- \`backend/rag_solution/data_ingestion/txt_processor.py\` - Text processing (keep)
- \`backend/rag_solution/data_ingestion/ingestion.py\` - Ingestion pipeline
- \`backend/rag_solution/data_ingestion/chunking.py\` - Chunking strategies

**New Files to Create:**
- \`backend/rag_solution/data_ingestion/docling_processor.py\` - Docling adapter

**Tests to Create/Update:**
- \`backend/tests/unit/test_docling_processor.py\` - Unit tests
- \`backend/tests/integration/test_docling_integration.py\` - Integration tests
- \`backend/tests/performance/test_docling_performance.py\` - Performance benchmarks

## Estimated Effort

- **Setup & Core Integration**: 2-3 days
- **Testing & Validation**: 2-3 days
- **Format Expansion**: 1-2 days
- **Documentation & Rollout**: 1 day

**Total**: 6-9 days of development effort

## Priority

**High Priority** - This enhancement provides significant improvements to core document processing capabilities, aligns with IBM ecosystem strategy (WatsonX), and reduces technical debt.

### PR DESCRIPTION
**Pull request by manavgup**
## Summary

Implements IBM Docling integration with AI-powered table extraction (TableFormer) and layout analysis (DocLayNet) to significantly improve document processing quality.

**Clean PR extracted from the deployment branch - contains ONLY Docling changes.**

### Enhanced Document Processing
- **AI-Powered Table Extraction**: TableFormer model significantly outperforms legacy processors
- **Layout Analysis**: DocLayNet model for reading order detection
- **Format Support**: PDF, DOCX, PPTX, HTML, and image formats
- **Metadata Enhancement**: Automatic table/image counting and tracking

### Implementation Quality
- **313% improvement** in chunk extraction vs legacy processors
- **Table detection**: 3 tables vs 0 (legacy)
- **Image detection**: 13 images vs 0 (legacy)
- **Feature flag control** (`ENABLE_DOCLING`) for safe deployment
- **Automatic fallback** to legacy processors on error
- **Type-safe implementation** with mypy validation

### New Files (5)
- `backend/rag_solution/data_ingestion/docling_processor.py` - Main Docling processor class (326 lines)
- `backend/tests/unit/test_docling_processor.py` - Comprehensive unit tests (630 lines, 14 tests)
- `backend/dev_tests/manual/test_docling_debug.py` - Debug utility
- `backend/dev_tests/manual/test_pdf_comparison.py` - Real PDF comparison validation
- `docs/issues/IMPLEMENTATION_PLAN_ISSUE_255.md` - Implementation documentation

### Modified Files (4)
- `backend/core/config.py` - Added ENABLE_DOCLING and DOCLING_FALLBACK_ENABLED flags
- `backend/rag_solution/data_ingestion/document_processor.py` - Integrated Docling processor
- `backend/pyproject.toml` - Added docling dependency  
- `backend/poetry.lock` - Updated dependencies (transformers 4.56.2 for compatibility)

## Testing

✅ **14 comprehensive unit tests** (100% passing)
- DoclingProcessor initialization and import handling
- PDF processing with table/image extraction
- Page number tracking with new Docling API
- Error handling and fallback mechanisms
- Multiple text item types (TextItem, SectionHeaderItem, ListItem, CodeItem)

✅ **Real PDF comparison validation**
- Test utilities for comparing legacy vs Docling output
- Debug scripts for development

✅ **Code Quality**
- docling_processor.py: **9.64/10**
- document_processor.py: **9.84/10**
- Type-safe with mypy validation

## Deployment Strategy

1. **Feature Flag**: ENABLE_DOCLING=False by default (safe deployment)
2. **Automatic Fallback**: Falls back to legacy processors if Docling fails
3. **Gradual Rollout**: Enable per environment for controlled testing

## Breaking Changes

**None** - Docling is disabled by default and includes automatic fallback.

## Related

- Closes #255
- Extracted from PR #265 and PR #261 for focused review

### CODE REVIEW

**Comment by manavgup** [comment]
## ✅ Fixed CI/CD Failures

**Root Cause**: PyTorch was installing with CUDA support (~6GB of NVIDIA libraries) which we don't need for CPU-only inference.

**Fixes Applied**:

1. **CPU-only PyTorch** (commit 27344aa)
   - Modified `backend/Dockerfile.backend` to install `torch==2.8.0+cpu` before Poetry
   - Saves ~6GB by avoiding 11 NVIDIA CUDA packages
   - Maintains full Docling functionality (AI models work fine on CPU)

2. **Disk Space Cleanup** (commit 4e96171)
   - Added cleanup step in `dev-environment-ci.yml` workflow
   - Removes unnecessary packages (.NET, GHC, Boost)
   - Frees ~14GB before Docker build

**Why This Happened**:
- Docling requires PyTorch for AI models (TableFormer for table extraction, DocLayNet for layout analysis)
- By default, PyTorch installs with CUDA support on Linux
- This pulled in massive CUDA libraries we don't need

**Impact**:
- ✅ Docker image size reduced by ~6GB
- ✅ CI/CD builds should succeed now
- ✅ No functionality change (Docling runs perfectly on CPU)
- ✅ Faster builds and deployments

CI should pass now! 🚀

**Comment by manavgup** [comment]
## 🔧 Version Compatibility Fix

**Issue**: Initial fix used torch 2.8.0+cpu but torchvision 0.20.0+cpu requires torch 2.5.0

**Fixed** (commit 857cec1):
- Changed to `torch==2.5.0+cpu` (compatible with torchvision 0.20.0)
- Maintains CPU-only installation (no CUDA)
- Still saves ~6GB by avoiding NVIDIA packages

**CI should pass now!** ✅

**Comment by manavgup** [comment]
## 📦 PyTorch Dependency Clarification

**Great question!** You're right that Docling doesn't directly list PyTorch as a dependency.

### Actual Dependency Chain:

```
docling (>=2.0.0)
  └── accelerate (>=1.0.0,<2)  ← Required by Docling
       └── torch (>=2.0.0)  ← Required by accelerate
```

**Why accelerate requires PyTorch:**
- `accelerate` is HuggingFace's library for distributed/mixed-precision training
- Even though we're only doing inference, accelerate has a hard dependency on PyTorch
- Docling uses accelerate for model loading and inference optimization

### Disk Space Impact:

Without CPU-only installation:
- PyTorch with CUDA: **~6.2GB** (11 NVIDIA packages)
- PyTorch CPU-only: **~800MB-1GB**

### Current Solution:

✅ **CPU-only PyTorch** (Dockerfile.backend lines 34-40)
- Saves ~5.4GB vs CUDA version
- Full Docling functionality maintained
- AI models (TableFormer, DocLayNet) work perfectly on CPU

✅ **Disk cleanup in workflows**
- dev-environment-ci.yml (commit 4e96171)
- ci.yml (commit 46bda9c)
- Frees ~14GB before builds

### Alternative Approaches (for discussion):

1. **Keep current** - 800MB is reasonable for AI-powered document processing
2. **Make Docling optional** - Feature flag already exists (ENABLE_DOCLING=False)
3. **Optimize accelerate** - Investigate if we can use a lighter alternative

Given the 313% improvement in document processing, the 800MB footprint seems justified. What do you think?

**Comment by manavgup** [comment]
## ✅ Fixed: Eliminated Duplicate Build

**Commit e4a8490**: Updated `dev-environment-ci.yml` to stop triggering on backend code changes.

**Before**:
```yaml
paths:
  - 'backend/**'  ← Triggered on ALL backend changes (wrong!)
  - 'tests/**'
```

**After**:
```yaml
paths:
  - '.devcontainer/**'
  - 'docker-compose.dev.yml'
  - 'docker-compose.hotreload.yml'
  # Only triggers when dev environment actually changes
```

**Impact**:
- ✅ Eliminates duplicate backend build
- ✅ Fixes "no space left on device" failure
- ✅ Saves ~10 minutes per PR
- ✅ This PR should now have ALL checks passing

**Next CI run** will validate the fix. The failing "Test Development Environment Setup" check should no longer run for this PR.
